File: src/main/java/cubesrv/CubeCosetsAdd.java
package cubesrv;

import static cubesrv.ThreadPoolHelper.*;
import java.util.concurrent.atomic.AtomicInteger;

public class CubeCosetsAdd {
    public static final int TWOPHASE_DEPTH1_CATCHFIRST_MAX = 5;
    public static final int TWOPHASE_DEPTH1_MULTI_MAX = 6;

    private static void addBGcosetsT(int threadNo, CubesReprByDepth cubesReprByDepth,
            CubeCosets bgCosets, int depth, AtomicInteger reprCubeCount) {
        int reprCubeCountT = 0;
        CubesReprAtDepth ccReprCubesC = cubesReprByDepth.getAt(depth);
        CornerPermReprCubes[] list = ccReprCubesC.ccpCubesList();
        for (int idx = 0; idx < list.length; idx++) {
            CornerPermReprCubes ccpCubes = list[idx];
            CubecornersPerm ccp = ccReprCubesC.getPermAt(idx);
            for (CornerOrientReprCubes ccoCubes : ccpCubes.ccoCubesList()) {
                CubecornerOrients cco = ccoCubes.getOrients();
                for (int reversed = 0; reversed < (cubesReprByDepth.isUseReverse() ? 2 : 1); reversed++) {
                    CubecornersPerm ccprev = (reversed != 0) ? ccp.reverse() : ccp;
                    CubecornerOrients ccorev = (reversed != 0) ? cco.reverse(ccp) : cco;
                    for (int symmetric = 0; symmetric <= 1; symmetric++) {
                        CubecornersPerm ccprevsymm = (symmetric != 0) ? ccprev.symmetric() : ccprev;
                        CubecornerOrients ccorevsymm = (symmetric != 0) ? ccorev.symmetric() : ccorev;
                        for (int td = 0; td < TransformDir.TCOUNT.ordinal(); td++) {
                            CubecornersPerm ccpT = ccprevsymm.transform(td);
                            CubecornerOrients ccoT = ccorevsymm.transform(ccprevsymm, td);
                            CubecornerOrients ccoReprBG = ccoT.representativeBG(ccpT);
                            int reprCOrientIdx = ccoReprBG.getOrientIdx();
                            if (reprCOrientIdx % THREAD_COUNT == threadNo) {
                                for (long edges : ccoCubes.edgeList()) {
                                    CubeEdges ce = new CubeEdges(edges);
                                    CubeEdges cerev = (reversed != 0) ? ce.reverse() : ce;
                                    CubeEdges cerevsymm = (symmetric != 0) ? cerev.symmetric() : cerev;
                                    CubeEdges ceT = cerevsymm.transform(td);
                                    CubeEdges ceReprBG = ceT.representativeBG();
                                    Cube c = new Cube(ccpT, ccoT, ceT);
                                    if (bgCosets.getAt(depth).addCube(reprCOrientIdx, ceReprBG, c))
                                        ++reprCubeCountT;
                                }
                            }
                        }
                    }
                }
            }
        }
        reprCubeCount.addAndGet(reprCubeCountT);
    }

    private static void addBGcosets(CubesReprByDepth cubesReprByDepth,
            CubeCosets bgCosets, int requestedDepth, Responder responder) {
        while (bgCosets.availCount() <= requestedDepth) {
            int depth = bgCosets.availCount();
            AtomicInteger reprCubeCount = new AtomicInteger(0);
            runInThreadPool(threadNo -> addBGcosetsT(threadNo, cubesReprByDepth, bgCosets, depth, reprCubeCount));
            responder.message("depth " + depth + " space repr cubes=" + reprCubeCount.get());
            bgCosets.incAvailCount();
        }
    }

    private final CubeCosets m_cubeCosets =
        new CubeCosets(Math.max(TWOPHASE_DEPTH1_CATCHFIRST_MAX, TWOPHASE_DEPTH1_MULTI_MAX) + 1);

    public CubeCosets getBGcosets(CubesReprByDepthAdd cubesReprByDepthAdd,
            int depth, Responder responder) {
        if (m_cubeCosets.availCount() <= depth) {
            CubesReprByDepth cubesReprByDepth = cubesReprByDepthAdd.getReprCubes(depth, responder);
            if (cubesReprByDepth == null) return null;
            addBGcosets(cubesReprByDepth, m_cubeCosets, depth, responder);
        }
        return m_cubeCosets;
    }
}

File: src/main/kotlin/cubesrv/CubeCosetsAdd.kt
package cubesrv

import java.util.concurrent.atomic.AtomicInteger

val TWOPHASE_DEPTH1_CATCHFIRST_MAX = 5
val TWOPHASE_DEPTH1_MULTI_MAX = 6

private fun addBGcosetsT(threadNo : Int,
        cubesReprByDepth : CubesReprByDepth, bgCosets : CubeCosets,
        depth : Int, reprCubeCount : AtomicInteger)
{
    var reprCubeCountT : Int = 0
    val ccReprCubesC : CubesReprAtDepth = cubesReprByDepth.getAt(depth)
    for(idx in 0 ..< ccReprCubesC.ccpCubesList().size) {
        val ccpCubes = ccReprCubesC.ccpCubesList()[idx]
        val ccp = ccReprCubesC.getPermAt(idx)
        for(ccoCubes in ccpCubes.ccoCubesList()) {
            val cco : cubecorner_orients = ccoCubes.getOrients()
            for(reversed in 0 ..< (if(cubesReprByDepth.isUseReverse()) 2 else 1)) {
                val ccprev : cubecorners_perm = if(reversed!=0) ccp.reverse() else ccp
                val ccorev : cubecorner_orients = if(reversed!=0) cco.reverse(ccp) else cco
                for(symmetric in 0..1) {
                    val ccprevsymm : cubecorners_perm = if(symmetric!=0) ccprev.symmetric() else ccprev
                    val ccorevsymm : cubecorner_orients = if(symmetric!=0) ccorev.symmetric() else ccorev
                    for(td in 0 ..< transform_dir.TCOUNT.ordinal) {
                        val ccpT : cubecorners_perm = ccprevsymm.transform(td)
                        val ccoT : cubecorner_orients = ccorevsymm.transform(ccprevsymm, td)
                        val ccoReprBG : cubecorner_orients = ccoT.representativeBG(ccpT)
                        val reprCOrientIdx : Int = ccoReprBG.getOrientIdx()

                        if( reprCOrientIdx % THREAD_COUNT == threadNo ) {
                            for(edges in ccoCubes.edgeList()) {
                                val ce = cubeedges(edges)
                                val cerev : cubeedges = if(reversed!=0) ce.reverse() else ce
                                val cerevsymm : cubeedges = if(symmetric!=0) cerev.symmetric() else cerev
                                val ceT : cubeedges = cerevsymm.transform(td)
                                val ceReprBG : cubeedges = ceT.representativeBG()
                                val c : cube = cube(ccp = ccpT, cco = ccoT, ce = ceT)
                                if( bgCosets.getAt(depth).addCube(reprCOrientIdx, ceReprBG, c) )
                                    ++reprCubeCountT
                            }
                        }
                    }
                }
            }
        }
    }
    reprCubeCount.addAndGet(reprCubeCountT)
}

private fun addBGcosets(cubesReprByDepth : CubesReprByDepth,
        bgCosets : CubeCosets, requestedDepth : Int, responder : Responder)
{
    while( bgCosets.availCount() <= requestedDepth ) {
        val depth : Int = bgCosets.availCount()
        val reprCubeCount = AtomicInteger(0)
        runInThreadPool() { threadNo ->
            addBGcosetsT(threadNo, cubesReprByDepth, bgCosets, depth, reprCubeCount)
        }
        responder.message("depth $depth space repr cubes=${reprCubeCount.get()}")
        bgCosets.incAvailCount()
    }
}

class CubeCosetsAdd {
    val m_cubeCosets = CubeCosets(maxOf(TWOPHASE_DEPTH1_CATCHFIRST_MAX, TWOPHASE_DEPTH1_MULTI_MAX)+1)

    /* Returns set of cosets filled up to at least the specified depth (inclusive).
     * The function triggers filling procedure when the depth is requested first time.
     * When the filling procedure is canceled, NULL is returned.
     */
    fun getBGcosets(cubesReprByDepthAdd : CubesReprByDepthAdd,
            depth : Int, responder : Responder) : CubeCosets?
    {
        if( m_cubeCosets.availCount() <= depth ) {
            val cubesReprByDepth : CubesReprByDepth? = cubesReprByDepthAdd.getReprCubes(depth, responder)
            if( cubesReprByDepth == null )
                return null
            addBGcosets(cubesReprByDepth, m_cubeCosets, depth, responder)
        }
        return m_cubeCosets
    }
}
